### ISSUE TITLE
TopK fails to analyze for a small LSTM model

### ISSUE THREAD

**Issue by alexander-camuto** [sonos/tract#1639]
A smallish LSTM fails to unify DIM types (`Impossible to unify TDim with I64`) for a TOPK node: 

Output from cli 
```
┏ 0 Source input_ids
┃   ━━━ batch_size,64,I64
┣┻ 10 Gather /embedding/Gather
┃   ━━━ batch_size,64,32,F32
┃┏ 1 Source attention_mask
┃┃   ━━━ batch_size,64,F32
┃┣ 11 Reduce<Sum> /ReduceSum
┃┃   ━━━ batch_size,F32
┃┣ 12 onnx.Cast /Cast
┃┣ 13 onnx.Cast /Cast_1
┃┃   ━━━ batch_size,TDim
┃┣┓ 
┃┃┣ 15 Shape /Shape
┃┃┃   ━━━ 1,TDim batch_size
┃┃┣┻ 17 Gather /Gather
┃┣╋ 18 Topk /TopK
┃┃┃   ━━━ ?,?
┃┃┃   ━━━ ?,?
┃┃┣ 19 onnx.Cast /Cast_2
┃┃┣┓ 
┗━┓┃ 
┃┃┣┻ 20 Gather /Gather_1
┃┃┃   ━━━ ?,?,?,F32
┃┃┣ 21 PermuteAxes /Transpose
┗━┓ 
┃┃┣ 23 onnx.Cast /Cast_3
┃┣━┓ 
┃┃┃┣ 22 Shape /Shape_1
┃┃┃┃   ━━━ ?,TDim
┃┃┃┣┻ 34 Gather /lstm/Gather
┃┃┃┃   ━━━ TDim
┃┃┃┣ 36 AddDims /lstm/Unsqueeze
┃┃┃┃   ━━━ 1,TDim
┃┃┃┣┻┻ 38 InferenceConcat /lstm/Concat
┃┃┃┃   ━━━ 3,TDim
┃┃┃┣ 39 ConstantOfShape /lstm/ConstantOfShape
┃┃┃┃   ━━━ 2,?,32,F32
 
┃┃ 
┃┣╋╋┻┻┻┻ 40 LSTM /lstm/LSTM
┃┃    ━━━ ?,2,?,32,F32
┃┃    ━━━ 2,?,32,F32
┃┃    ━━━ 2,?,32,F32
┃┣ 41 PermuteAxes /lstm/Transpose
┃┃   ━━━ ?,?,?,?,F32
┃┣┻ 43 Reshape /lstm/Reshape
┃┃   ━━━ ?,?,?,F32
┃┣ 44 PermuteAxes /Transpose_1
┣━┓ 
┃┃┣ 30 Shape /Shape_3
┃┃┃   ━━━ ?,TDim
┃┃┣ 31 ConstantOfShape /ConstantOfShape
┃┃┃   ━━━ ..,I64
┣━━┓ 
┃┃┃┣ 24 Shape /Shape_2
┃┃┃┃   ━━━ ?,TDim
┃┃┃┣ 25 Reduce<Prod> /ReduceProd
┃┃┃┃   ━━━ TDim
┃┃┃┣ 27 onnx.Cast /Cast_4
┃┃┃┣┻┻ 29 Range /Range
┃┃┃┃   ━━━ ?,TDim
┗━┓┃ 
┃┣┻┻ 32 ScatterElements /ScatterElements
┃┃   ━━━ ?,?
┣┻ 45 Gather /Gather_2
┣┻ 46 Gather /Gather_3
┃   ━━━ ?,64,F32
┣┻┻ 47 Gemm /layers/layers.0/Gemm
┃   ━━━ ?,32,F32
┣ 48 Clip /layers/layers.1/Relu
┣┻┻ 49 Gemm /layers/layers.3/Gemm
    ━━━ ?,20,F32
[2025-02-06T18:37:57.164582000Z ERROR tract] Error at stage "analyse"
    
    Caused by:
        0: ModelBuildingError
        1: Failed analyse for node #18 "/TopK" Topk
        2: Infering facts
        3: Applying rule inputs[1].datum_type == I64
        4: Impossible to unify TDim with I64.

```

Model 

[network.onnx.zip](https://github.com/user-attachments/files/18694692/network.onnx.zip)

To reproduce 

```
tract network.onnx 

```

**Comment by ZAZPRO** [comment]
Same problem with different model:

```
Error at stage "analyse"
    
    Caused by:
        0: ModelBuildingError
        1: Failed analyse for node #1268 "/text_encoder_1/TopK_5" Topk
        2: Infering facts
        3: Applying rule inputs[1].datum_type == I64
        4: Impossible to unify TDim with I64.
```

**Comment by kali** [comment]
@ZAZPRO thanks for putting this issue back on the table.
@alexander-camuto sorry i forgot about this one while you took the trouble to prepare a very exploitable test-case.

**Comment by kali** [comment]
So there are multiple issues triggered by this model. First there is a some weakness in the support of symbolic dimension in the TopK and the ScatterElements operators. These one are fixed in the linked PR.

Next comes the onnx LSTM translator with the same problem. It gets a symbolic seq_len. I don't think this will be a big problem to fix, stay tuned.

**Comment by kali** [comment]
Netron view of the ONNX model

<img width="514" height="1397" alt="Image" src="https://github.com/user-attachments/assets/db9a421b-8922-4a05-b846-19f775153a1f" />

All the cast around the TopK node are problematic, because they flatten the TDim before reaching the LSTM, so the symbolic dimension is lost before reaching the LSTM seq_len.
